sdk/identity/identity/src/msal/errors.ts

```diff
@@ -20,6 +20,6 @@ export class AuthenticationRequired extends CredentialUnavailable {
     message?: string
   ) {
     super(message);
-    this.name = "AuthenticationRequired";
+    this.name = "AuthenticationRequiredError";
```

Comment:
Looks like the constructor of this class takes in scopes and getTokenOptions. How do these parameters get used?

Reply:
They're used to debug, they're public properties 🌞

Reply:
It is for the user to use for debugging or for us? How is it that none of the other error classes need these for debugging?

Reply:
AuthenticationRequiredError is the only error that is exclusively related to getToken at the moment, so we want to expose the getToken parameters (scopes and options) for users to have more context of why things failed.

Reply:
Can you point us to the PR that made this change? Maybe I am missing something... 

Usually if we want users to gain context from an error, we add the details to the error message and make use of the error code. This is a new pattern I am seeing where we are adding custom fields to the Error class. Have we run this pattern for Error classes with the architects?

Looking at the changelog for `AuthenticationRequiredError`, looks like we expect users to call `authenticate()` when they encounter this to do the authentication manually. Is that why we are exposing the `scopes` and the `options`, because the `authenticate()` method would need this information?

Reply:
We exposed these properties as part of the tasks we did after we reviewed the public API differences across languages. PR: https://github.com/Azure/azure-sdk-for-js/pull/14629 which fixed issue: https://github.com/Azure/azure-sdk-for-js/issues/14585

Java has similar code: https://github.com/Azure/azure-sdk-for-java/blame/master/sdk/identity/azure-identity/src/main/java/com/azure/identity/AuthenticationRequiredException.java

I can't find the reasoning right away, but what you say makes sense.

"AutenticationRequired" fails but allows you to authenticate with the same parameters. The exposition of parameters is something that made sense to me already, this frame makes it evidently useful 👍

Reply:
There's an old note on Python regarding how `AuthenticationRequiredError.claims` can be used in a subsequent authenticate() call: https://github.com/Azure/azure-sdk-for-python/blob/master/sdk/identity/azure-identity/CHANGELOG.md#added-1

Seems like part of the underlying meaning has been lost since this error was added, then removed, then re-introduced in other languages (case for Python: [link](https://github.com/Azure/azure-sdk-for-python/blob/master/sdk/identity/azure-identity/CHANGELOG.md#150-2020-11-11), case for Java: [link](https://github.com/Azure/azure-sdk-for-java/blob/master/sdk/identity/azure-identity/CHANGELOG.md#110-2020-08-10)). Meanwhile in JS we were considerably far behind.

Reply:
> AuthenticationRequiredError is the only error that is exclusively related to getToken at the moment, so we want to expose the getToken parameters (scopes and options) for users to have more context of why things failed.

As @ramya-rao-a surmised the reason why `AuthenticationRequireError` contains all the information from the request to `getToken` is because users are expected to handle this error and recover by calling `authenticate` to perform the user interaction. They need to specify the same parameters to `authenticate` as was originally requested from `getToken` so the proper interaction is performed.

> Have we run this pattern for Error classes with the architects?

Yes, this API, along with the champion scenario involving it have been reviewed by the architecture board.